packages/nimble-components/src/table-column/mixins/custom-sort-order.ts

```diff
@@ -0,0 +1,93 @@
+import { Notifier, Observable, attr } from '@microsoft/fast-element';
+import type { TableColumn } from '../base';
+import { TableColumnSortOperation } from '../base/types';
+import type { ColumnValidator } from '../base/models/column-validator';
+
+// Pick just the relevant properties the mixin depends on (typescript complains if the mixin declares private / protected base exports)
+type CustomSortOrderTableColumn<
+    TColumnValidator extends ColumnValidator<['invalidCustomSortWithGrouping']>
+> = Pick<TableColumn<unknown, TColumnValidator>, 'columnInternals'>;
+// prettier-ignore
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type CustomSortOrderTableColumnConstructor<TColumnValidator extends ColumnValidator<['invalidCustomSortWithGrouping']>> = abstract new (...args: any[]) => CustomSortOrderTableColumn<TColumnValidator>;
+
+// As the returned class is internal to the function, we can't write a signature that uses is directly, so rely on inference
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
+export function mixinCustomSortOrderColumnAPI<
+    TBase extends CustomSortOrderTableColumnConstructor<TColumnValidator>,
+    TColumnValidator extends ColumnValidator<['invalidCustomSortWithGrouping']>
+>(base: TBase) {
+    /**
+     * The Mixin that provides a concrete column with the API to allow sorting
+     * by a field other than the default for that column.
+     */
+    abstract class CustomSortOrderColumn extends base {
+        public sortByFieldName?: string;
+
+        /** @internal */
+        public notifier?: Notifier;
+
+        /** @internal */
+        public sortByFieldNameChanged(): void {
+            this.updateOperandDataRecordFieldName();
+
+            if (typeof this.sortByFieldName === 'string' && !this.notifier) {
+                this.notifier = Observable.getNotifier(this.columnInternals);
+                this.notifier.subscribe(this);
+                this.updateCustomColumnSortingValidity();
+            } else {
+                this.notifier?.unsubscribe(this);
+                this.notifier = undefined;
+                this.updateCustomColumnSortingValidity();
+            }
+        }
+
+        /** @internal */
+        public updateOperandDataRecordFieldName(): void {
+            if (typeof this.sortByFieldName === 'string') {
+                this.columnInternals.operandDataRecordFieldName = this.sortByFieldName;
+                this.columnInternals.sortOperation = TableColumnSortOperation.basic;
+            } else {
+                this.columnInternals.operandDataRecordFieldName = this.getDefaultSortFieldName();
+                this.columnInternals.sortOperation = this.getDefaultSortOperation();
+            }
+        }
+
+        /** @internal */
+        public handleChange(_source: unknown, args: unknown): void {
```

Comment:
hmm...interesting question. The tricky thing here is that if `this.updateOperandDataRecordFieldName()` is not called, then the column class needs to set `this.columnInternals.operandDataRecordFieldName = <newFieldName>;`, and that has nothing to do with custom sorting behavior. Therefore, it doesn't really feel like a private implementation detail of the mixin that should be completely hidden away from a column class, so I think it would make more sense to leave the code as-is. However, I could also see the argument that it's better to have the code in a single location, and column tests would (hopefully) catch any problems if the internals of the mixin changed such that the `this.columnInternals.operandDataRecordFieldName` wasn't updated at the correct time.

Reply:
Ah, I see what you're getting at. So if we wanted to go in the direction I suggested, we would really want to move `getDefaultSortFieldName`, `getDefaultSortOperation`, and a trivial implementation of `updateOperandDataRecordFieldName` to the `SortableColumn` mixin class, then make `CustomSortOrderColumn` extend it (actually, we can't extend it, but instead we add those member names to the `Pick<> at the top of the file?)`. Column types would then no longer call `this.updateOperandDataRecordFieldName()` or set `this.columnInternals.operandDataRecordFieldName = <newFieldName>`. However, they would all have to implement those two abstract members (or read-only properties). I think I like that.

The one reamaining loose end I can think of is that if a column that is groupable but not sortable would still have to set `operandDataRecordFieldName` itself. That doesn't seem too bad to me, especially given how (presumably) rare it would be to support only grouping without sorting.

All that said, there's nothing wrong with the code as it is now, so it's just a suggestion.

Reply:
What I realized reading your comment is that we don't enforce that the `CustomSortOrderColumn` mixin is only used by columns that also use the `SortableColumn` mixin. I don't think that's a huge deal, especially since I'm not sure how we'd go about enforcing that two mixins were used together.

But, that being said, I'm not sure how we'd `Pick` the fields you mention from the `CustomSortOrderColumn` since we can only `Pick` things from the base `TableColumn` class.

Reply:
> we can only Pick things from the base TableColumn class.

Oh, right. Hmm. How important is it to keep sortability and custom-sortability separate? I can't really think of a reason why we'd need to disallow custom sorting for an otherwise sortable column. If there isn't one, we could just merge the two mixins.

This is sort of tangential, but when thinking of merging the mixins, it occurred to me that `sortByFieldName` might be better named `customSortFieldName`. Otherwise, it's kind of unintuitive that, for example, `sortByFieldName` is unset, and yet there is a field that you're sorting by.

Reply:
> I can't really think of a reason why we'd need to disallow custom sorting for an otherwise sortable column.

Our philosophy tends to be that we should try to limit what a client can do unless we have a use case for it. I don't think we have any reason to allow custom sorting for numeric columns (duration, date, number-text). Those columns are all sortable, but I think a custom sort order would be pretty nonsensible.

As far as the name of the property, I don't have a strong opinion either way. @rajsite and @jattasNI, any opinions on the name `sortByFieldName` vs `customSortFieldName`.

Reply:
I don't think `customSortFieldName` is significantly better than `sortByFieldName` so I vote to keep the decision unchanged from what was in the HLD.